single-node-refactor/src/Solvers/SGH_solver/src/contact.cpp

```diff
@@ -0,0 +1,463 @@
+    REDUCE_MAX_CLASS(i, 0, contact_patch_t::num_nodes_in_patch*num_contact_patches, local_y_max, {
+        if (local_y_max < points(1, i))
+        {
+            local_y_max = points(1, i);
+        }
+    }, y_max);
+    double local_z_max = 0.0;
+    REDUCE_MAX_CLASS(i, 0, contact_patch_t::num_nodes_in_patch*num_contact_patches, local_z_max, {
+        if (local_z_max < points(2, i))
+        {
+            local_z_max = points(2, i);
+        }
+    }, z_max);
+    double local_x_min = 0.0;
+    REDUCE_MIN_CLASS(i, 0, contact_patch_t::num_nodes_in_patch*num_contact_patches, local_x_min, {
+        if (local_x_min > points(0, i))
+        {
+            local_x_min = points(0, i);
+        }
+    }, x_min);
+    double local_y_min = 0.0;
+    REDUCE_MIN_CLASS(i, 0, contact_patch_t::num_nodes_in_patch*num_contact_patches, local_y_min, {
+        if (local_y_min > points(1, i))
+        {
+            local_y_min = points(1, i);
+        }
+    }, y_min);
+    double local_z_min = 0.0;
+    REDUCE_MIN_CLASS(i, 0, contact_patch_t::num_nodes_in_patch*num_contact_patches, local_z_min, {
+        if (local_z_min > points(2, i))
+        {
+            local_z_min = points(2, i);
+        }
+    }, z_min);
+    double local_vx_max = 0.0;
+    REDUCE_MAX_CLASS(i, 0, contact_patch_t::num_nodes_in_patch*num_contact_patches, local_vx_max, {
+        if (local_vx_max < velocities(0, i))
+        {
+            local_vx_max = velocities(0, i);
+        }
+    }, vx_max);
+    double local_vy_max = 0.0;
+    REDUCE_MAX_CLASS(i, 0, contact_patch_t::num_nodes_in_patch*num_contact_patches, local_vy_max, {
+        if (local_vy_max < velocities(1, i))
+        {
+            local_vy_max = velocities(1, i);
+        }
+    }, vy_max);
+    double local_vz_max = 0.0;
+    REDUCE_MAX_CLASS(i, 0, contact_patch_t::num_nodes_in_patch*num_contact_patches, local_vz_max, {
+        if (local_vz_max < velocities(2, i))
+        {
+            local_vz_max = velocities(2, i);
+        }
+    }, vz_max);
+    double local_ax_max = 0.0;
+    REDUCE_MAX_CLASS(i, 0, contact_patch_t::num_nodes_in_patch*num_contact_patches, local_ax_max, {
+        if (local_ax_max < accelerations(0, i))
+        {
+            local_ax_max = accelerations(0, i);
+        }
+    }, ax_max);
+    double local_ay_max = 0.0;
+    REDUCE_MAX_CLASS(i, 0, contact_patch_t::num_nodes_in_patch*num_contact_patches, local_ay_max, {
+        if (local_ay_max < accelerations(1, i))
+        {
+            local_ay_max = accelerations(1, i);
+        }
+    }, ay_max);
+    double local_az_max = 0.0;
+    REDUCE_MAX_CLASS(i, 0, contact_patch_t::num_nodes_in_patch*num_contact_patches, local_az_max, {
+        if (local_az_max < accelerations(2, i))
+        {
+            local_az_max = accelerations(2, i);
+        }
+    }, az_max);
+    Kokkos::fence();
+
+    // If the max velocity is zero, then we want to set it to a small value. The max velocity and acceleration are used
```

Comment:
If the maximum velocity is zero, then there is no motion, so a collision cannot happen.  I may be wrong, but if that's the case we may want to just exit out of collision here if a collision is impossible.